modules/streamloader/src/main/scala/com/snowplowanalytics/snowplow/storage/bigquery/streamloader/Resources.scala

```diff
@@ -0,0 +1,242 @@
+    metrics: Metrics[F],
+    sinkSettingsBad: SinkSettings.Bad
+  ): Resource[F, Pipe[F, StreamBadRow[F], Unit]] =
+    mkProducer[F, BadRow](
+      projectId,
+      topic,
+      sinkSettingsBad.producerBatchSize,
+      sinkSettingsBad.producerDelayThreshold
+    ).map { p =>
+      _.parEvalMapUnordered(maxConcurrency) { badRow =>
+        p.produce(badRow.row) *> badRow.ack *> metrics.badCount
+      }
+    }
+
+  // Does not ack the event -- it still needs to end up in one of the other targets
+  private def mkTypeSink[F[_]: Concurrent](
+    projectId: String,
+    topic: String,
+    sinkSettingsTypes: SinkSettings.Types
+  ): Resource[F, Pipe[F, Set[ShreddedType], Unit]] =
+    mkProducer[F, Set[ShreddedType]](
+      projectId,
+      topic,
+      sinkSettingsTypes.producerBatchSize,
+      sinkSettingsTypes.producerDelayThreshold
+    ).map { p =>
+      _.parEvalMapUnordered(sinkSettingsTypes.sinkConcurrency) { types =>
+        p.produce(types).void
+      }
+    }
+
+  /** Write good events through a BigQuery sink and pipe observed types through a Pub/Sub sink.
+    * Acks the events.
+    *
+    * The input Stream of events is transformed into a Stream of batches of events (List[StreamLoaderRow]).
+    * Each batch respects the Streaming APIs recommended number of records per request.
+    * Each BQ write request also has a size limit of 10MB. In case a batch's total size exceeds this limit, we use
+    * as many events as possible (while staying below it) in the write request, and re-insert the remaining events
+    * back into the original Stream of events.
+    *
+    * @param bufferOverflowQueue A Queue in which put events that are left over from a batch after taking as many as
+    *                            possible without breaching the streaming inserts request size limit
+    */
+  private def mkGoodSink[F[_]: Concurrent: ContextShift: Timer](
+    blocker: Blocker,
+    good: Output.BigQuery,
+    bigQuery: BigQuery,
+    producer: PubsubProducer[F, FailedInsert],
+    metrics: Metrics[F],
+    typeSink: Pipe[F, Set[ShreddedType], Unit],
+    bufferOverflowQueue: Queue[F, StreamLoaderRow[F]],
+    sinkSettingsGood: SinkSettings.Good,
+    sinkSettingsTypes: SinkSettings.Types
+  ): Pipe[F, StreamLoaderRow[F], Unit] = {
+    val goodPipe: Pipe[F, StreamLoaderRow[F], Unit] = {
+      _.merge(bufferOverflowQueue.dequeue)
+        .through(batch(sinkSettingsGood.bqWriteRequestThreshold, sinkSettingsGood.bqWriteRequestTimeout))
+        .parEvalMapUnordered(sinkSettingsGood.sinkConcurrency) { slrows =>
+          val (remaining, toInsert) = splitBy(slrows, getSize[F], sinkSettingsGood.bqWriteRequestSizeLimit)
+          remaining.traverse_(bufferOverflowQueue.enqueue1) *>
+            blocker.blockOn(
+              Bigquery.insert(producer, metrics, toInsert.map(_.row))(Bigquery.mkInsert(good, bigQuery)) *> toInsert
+                .traverse_(_.ack)
+            )
+        }
+    }
+
+    _.observe(goodPipe)
```

Comment:
Aaaaggghhh - DANGER HERE.

Here you have a `parEvalMapUnordered` inside of a `observer` block.  This is exactly the combination that caused the bug in enrich, where we ended up acking messages before they had been processed.

Sorry for the bad news.  I haven't thought through what are the consequences in this particular case, but I think this needs some attention.

Reply:
Thanks for the headsup on this @istreeter. I used your demo example from before to convince myself (I think!) that the good pipe is safe because both sinking and acking happen within the same parEvalMap call. The types aggregation and sink also seems to work. I'm adding a commit with this so it can be tested easily.

```
package com.snowplowanalytics.snowplow.storage.bigquery.demo

import cats.effect.{ExitCode, IO, IOApp}
import fs2.{Pipe, Stream}

import scala.concurrent.duration.{DurationLong, FiniteDuration}

object Main extends IOApp {

  def ack(i: Int): IO[Unit] =
    IO.delay(println(s"acking message $i"))

  def sink: Pipe[IO, Int, Unit] =
    _.parEvalMapUnordered(1)(i => IO.sleep(1.seconds) *> IO.delay(println(s"sinking message $i")) *> ack(i))

  def aggregate(n: Int, t: FiniteDuration): Pipe[IO, Int, Set[Int]] =
    _.groupWithin(n, t).map(chunk => chunk.toList.toSet)

  def sink2: Pipe[IO, Set[Int], Unit] =
    _.parEvalMapUnordered(1)(i => IO.sleep(1.seconds) *> IO.delay(println(s"sinking2 message $i")))

  def run(args: List[String]): IO[ExitCode] =
    Stream(1, 1, 2, 2, 1, 3, 1,
      2).covary[IO].observe(sink).through(aggregate(2, 1.seconds)).through(sink2).compile.drain.as(ExitCode.Success)
}

```

```
sbt 'project demo' run

sinking message 1
acking message 1
sinking message 1
acking message 1
sinking message 2
acking message 2
sinking message 2
acking message 2
sinking message 1
acking message 1
sinking message 3
acking message 3
sinking message 1
acking message 1
sinking message 2
acking message 2
sinking2 message Set(1)
sinking2 message Set(2)
sinking2 message Set(1, 3)
sinking2 message Set(1, 2)
```

Reply:
I think you are right about sinking/acking.  An ack will always come after sinking the message, because you have them in the same parEvalMap - as you said.  That's good.

My concern is about the order of the typeSink and the goodSink.  I think the danger is that a ShreddedType could get emitted to the type sink before the corresponding StreamLoaderRow is emitted to the good sink.  If I'm right, then you have to imagine the situation where the ShreddedType is emitted, and the StreamLoaderRow fails to be emitted (maybe because of resource unavailable, or app crashes).  Maybe that's not a problem though if that happens?

Reply:
I think I don't see a downside to that. If the type was known to Mutator, it will be ignored. If it wasn't, then the sooner Mutator sees it the better.

However, from this limited experiment above, it looks like the type sink (`sink2`) always comes after `sink *> ack` anyway. Do you have strong concerns that the demo is flawed and shouldn't be trusted?

Reply:
I trust you there is no downside, and your reasoning makes sense.  So no change is necessary.

I adapted your demo -- I changed `1.seconds` to `10.seconds` in your `sink` function, so it is slower than `sink2`. Now it perfectly demonstrates the problem:

```
sinking message 1
acking message 1
sinking message 1
acking message 1
sinking message 2
acking message 2
sinking message 2
acking message 2
sinking message 1
acking message 1
sinking message 3
acking message 3
sinking message 1
acking message 1
sinking2 message Set(1)
sinking2 message Set(2)
sinking2 message Set(1, 3)
sinking2 message Set(1, 2)
sinking message 2
acking message 2
```

Maybe just add a code comment, just in case someone ever tries to extend the code and falls into this trap.